### PR TITLE
[minor] overload from_unixtime func to have optional timezone parameter

### PR DESCRIPTION
**Pull request by buraksenn**
## Which issue does this PR close?
Not closes but discussed in #12892

## Rationale for this change
from_unixtime is not aware of timezone so this PR makes it have optional timezone parameter

## What changes are included in this PR?
code,test,doc change for overloaded from_unixtime

## Are these changes tested?
yes added tests

## Are there any user-facing changes?
yes

### CODE REVIEW

**Review by jonathanc-n — APPROVED** [review]
This LGTM! Thank you very much

File: docs/source/user-guide/sql/scalar_functions_new.md
Comment (Omega359) [inline docs/source/user-guide/sql/scalar_functions_new.md:2006]: The common way to denote optional parameters is with [, optional_param]
```suggestion
from_unixtime(expression[, timezone])
```

File: datafusion/functions/src/datetime/from_unixtime.rs
Comment (Omega359) [inline datafusion/functions/src/datetime/from_unixtime.rs:130]: This probably should be just `.with_stardard_argument("expression", None)`

File: datafusion/functions/src/datetime/from_unixtime.rs
Comment (Omega359) [inline datafusion/functions/src/datetime/from_unixtime.rs:143]: These tests are great to have but it would be very nice to have a few additional ones in timestamps.slt as well

File: datafusion/functions/src/datetime/from_unixtime.rs
Comment (Omega359) [inline datafusion/functions/src/datetime/from_unixtime.rs:127]: A SQL example would be a very nice addition to have. The way I typically do that is to [build the CLI](https://datafusion.apache.org/user-guide/cli/index.html) with docker, run it and execute the desired sql in the cli and copy/paste it into the docs here.

**Comment by jonathanc-n** [comment]
I can probably finish the rest of the issue in another pr after this is merged

**Comment by alamb** [comment]
I took the liberty of merging this PR up from main to resolve a conflict

File: datafusion/functions/src/datetime/from_unixtime.rs
Comment (alamb) [reply datafusion/functions/src/datetime/from_unixtime.rs:127]: I actually did this and found a bug: 

```sql
> select from_unixtime(1730592000, 'America/New_York');
Internal error: Failed due to a difference in schemas, original schema: DFSchema { inner: Schema { fields: [Field { name: "from_unixtime(ts_data.ts / Int64(1000000000),Utf8(\"America/New_York\"))", data_type: Timestamp(Second, Some("Utf8")), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "from_unixtime(Int64(1730592000),Utf8(\"America/New_York\"))", data_type: Timestamp(Second, Some("Utf8")), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "arrow_typeof(from_unixtime(Int64(1730592000),Utf8(\"America/New_York\")))", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, field_qualifiers: [None, None, None], functional_dependencies: FunctionalDependencies { deps: [] } }, new schema: DFSchema { inner: Schema { fields: [Field { name: "from_unixtime(ts_data.ts / Int64(1000000000),Utf8(\"America/New_York\"))", data_type: Timestamp(Second, Some("Utf8")), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "from_unixtime(Int64(1730592000),Utf8(\"America/New_York\"))", data_type: Timestamp(Second, Some("America/New_York")), nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "arrow_typeof(from_unixtime(Int64(1730592000),Utf8(\"America/New_York\")))", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, field_qualifiers: [None, None, None], functional_dependencies: FunctionalDependencies { deps: [] } }.
```

Comment (alamb) [reply datafusion/functions/src/datetime/from_unixtime.rs:127]: I think the difference is that you need to take the *value* of the timezone argument, not just the type
 
You can fix this by using `return_type_from_exprs` instead of `return_type`:

https://github.com/apache/datafusion/blob/b7f4db4b56e945944c0448f84c61b6f3b86728f8/datafusion/expr/src/udf.rs#L451

**Review by alamb — COMMENTED** [review]
Thanks @buraksenn -- I am marking this PR as draft as it is no longer waiting on review (and I am trying to clear the review queue)

I did merge again from main to fix the compile error

File: datafusion/functions/src/datetime/from_unixtime.rs
Comment (buraksenn) [reply datafusion/functions/src/datetime/from_unixtime.rs:155]: added

File: datafusion/functions/src/datetime/from_unixtime.rs
Comment (buraksenn) [reply datafusion/functions/src/datetime/from_unixtime.rs]: Thanks, @alamb I could not find time before today I was debugging the issue. Fixed it with what you've said and added an SQL example

**Comment by buraksenn** [comment]
> Thanks @buraksenn -- I am marking this PR as draft as it is no longer waiting on review (and I am trying to clear the review queue)
> 
> I did merge again from main to fix the compile error

Thanks for the review. I've fixed the issue and added test cases to timestamps.slt. I'm converting it to Ready for review again

File: datafusion/functions/src/datetime/from_unixtime.rs
Comment (berkaysynnada) [inline datafusion/functions/src/datetime/from_unixtime.rs:88]: Can you direct to `return_type_from_exprs`

**Review by berkaysynnada — APPROVED** [review]
LGTM, thank you @buraksenn

File: datafusion/functions/src/datetime/from_unixtime.rs
Comment (buraksenn) [reply datafusion/functions/src/datetime/from_unixtime.rs:88]: In the docs it says that if `return_type_from_expr` is implemented we should return internal error so that's why I did this.

https://github.com/buraksenn/datafusion/blob/b7f4db4b56e945944c0448f84c61b6f3b86728f8/datafusion/expr/src/udf.rs#L419-L421: 
```
    /// What [`DataType`] will be returned by this function, given the types of
    /// the arguments.
    ///
    /// # Notes
    ///
    /// If you provide an implementation for [`Self::return_type_from_exprs`],
    /// DataFusion will not call `return_type` (this function). In this case it
    /// is recommended to return [`DataFusionError::Internal`].
```

Comment (buraksenn) [reply datafusion/functions/src/datetime/from_unixtime.rs]: aa got it. done

**Pull request merged**
